core/drivers/crypto/caam/caam_jr.c

```diff
@@ -0,0 +1,781 @@
+// SPDX-License-Identifier: BSD-2-Clause
+/*
+ * Copyright 2018-2019 NXP
+ *
+ * Brief   CAAM Job Rings manager.
+ *         Implementation of functions to enqueue/dequeue CAAM Job Descriptor
+ */
+
+/* Global includes */
+#include <kernel/interrupt.h>
+#include <kernel/panic.h>
+#include <kernel/pm.h>
+#include <kernel/spinlock.h>
+#include <mm/core_memprot.h>
+#include <tee/cache.h>
+
+/* Local includes */
+#include <caam_common.h>
+#include <caam_io.h>
+#include <caam_jr.h>
+#include <caam_hal_jr.h>
+#include <caam_rng.h>
+#include <caam_utils_delay.h>
+#include <caam_utils_mem.h>
+
+/*
+ * Enable the interrupt mode or not
+ */
+#ifdef CFG_CRYPTO_DRIVER
+//#define IT_MODE
+#endif
+
+/*
+ * Definition of input ring object
+ */
+struct inring_entry {
+	descEntry_t desc; /* Physical address of the descriptor */
+};
+
+/*
+ * Definition of output ring object
+ */
+struct __attribute__((__packed__)) outring_entry {
+	descEntry_t  desc;   /* Physical address of the descriptor */
+	descStatus_t status; /* Status of the executed job */
+};
+
+/*
+ * Job Free define
+ */
+#define JR_JOB_FREE	0
+
+/*
+ * Caller information context object
+ */
+struct caller_info {
+	struct jr_jobctx *jobctx; /* Caller job context object */
+	uint32_t         jobid;   /* Current Job Id */
+	descEntry_t      pdesc;   /* Physical address of the descriptor */
+};
+
+/*
+ * Job Ring module private data
+ */
+struct jr_privdata {
+	vaddr_t baseaddr;               /* Job Ring base address */
+
+	vaddr_t ctrladdr;               /* CAAM virtual base address */
+	paddr_t jroffset;               /* Job Ring address offset */
+	uint64_t paddr_inrings;         /* Physical address of input queue */
```

Comment:
Why aren't these `paddr_t` instead?

Reply:
This is the HW CAAM configuration registers that are 64 bits coded.

Reply:
Yes to get the physical address but HW is 2x32 bits registers for this. I'll add comments.